File: src/main/java/bernard/tatin/ProcFS/ProcessID.java
package bernard.tatin;


public class ProcessID {
    final private static ProcessID ourInstance = new ProcessID();
    private long pid = -1;

    private long innerGetPID() {
        if (pid == -1) {
            String processName =
                    java.lang.management.ManagementFactory.getRuntimeMXBean().getName();
            pid = Long.parseLong(processName.split("@")[0]);
        }
        return pid;
    }

    public static long getPID() {
        return ProcessID.ourInstance.innerGetPID();
    }


    private ProcessID() {
    }
}

File: kotlin-version/src/main/kotlin/bernard/tatin/WorkInProgress/processid/ProcessID.kt
package bernard.tatin.ProcFS.processid


object ProcessID {
    private var pid: Long = badPID()

    fun badPID() : Long {
        return -1
    }
    fun getPID(): Long {
        if (pid == badPID()) {
            val processName = java.lang.management.ManagementFactory.getRuntimeMXBean().getName()
            pid = processName.split("@")[0].toLong()
        }
        return pid
    }
}
